Mathlib/Combinatorics/SimpleGraph/CompleteMultipartite.lean

```diff
@@ -0,0 +1,152 @@
+import Mathlib.Combinatorics.SimpleGraph.Hasse
+
+/-!
+# Complete Multipartite Graphs
+
+A graph is complete multipartite iff non-adjacency is transitive.
+
+## Main declarations
+
+* `SimpleGraph.IsCompleteMultipartite`: predicate for a graph to be complete-multi-partite.
+
+* `SimpleGraph.IsCompleteMultipartite.setoid`: the `Setoid` given by non-adjacency.
+
+* `SimpleGraph.IsCompleteMultipartite.iso`: the graph isomorphism from a graph that
+  `IsCompleteMultipartite` to the corresponding `completeMultipartiteGraph`.
+
+* `SimpleGraph.IsPathGraph3Compl`: predicate for three vertices to be a witness to
+   non-complete-multi-partite-ness of a graph G. (The name refers to the fact that the three
+   vertices form the complement of `pathGraph 3`.)
+-/
+
+universe u
+namespace SimpleGraph
+variable {α : Type u}
+
+/-- `G` is `IsCompleteMultipartite` iff non-adjacency is transitive -/
+def IsCompleteMultipartite (G : SimpleGraph α) : Prop := Transitive (¬ G.Adj · ·)
+
+variable {G : SimpleGraph α}
+/-- The setoid given by non-adjacency -/
+def IsCompleteMultipartite.setoid (h : G.IsCompleteMultipartite) : Setoid α :=
+    ⟨(¬ G.Adj · ·), ⟨G.loopless, fun h' ↦ by rwa [adj_comm] at h', fun h1 h2 ↦ h h1 h2⟩⟩
+
+lemma completeMultipartiteGraph.isCompleteMultipartite {ι : Type*} (V : ι → Type*) :
+    (completeMultipartiteGraph V).IsCompleteMultipartite := by
+  intro
+  aesop
+
+/-- The graph isomorphism from a graph `G` that `IsCompleteMultipartite` to the corresponding
+`completeMultipartiteGraph` -/
+def IsCompleteMultipartite.iso (h : G.IsCompleteMultipartite) :
+    G ≃g completeMultipartiteGraph (fun (c : Quotient h.setoid) ↦ {x // h.setoid.r c.out x}) where
+  toFun := fun x ↦ ⟨_, ⟨_, Quotient.mk_out x⟩⟩
+  invFun := fun ⟨_, x⟩ ↦  x.1
+  left_inv := fun _ ↦ rfl
+  right_inv := fun ⟨_, x⟩ ↦ Sigma.subtype_ext (Quotient.mk_eq_iff_out.2 <| h.setoid.symm x.2) rfl
+  map_rel_iff' := by
+    simp_rw [Equiv.coe_fn_mk, comap_adj, top_adj, ne_eq, Quotient.eq]
+    intros
+    change ¬¬ G.Adj _ _ ↔ _
+    rw [not_not]
+
+lemma isCompleteMultipartite_iff : G.IsCompleteMultipartite ↔ ∃ (ι : Type u) (V : ι → Type u)
+    (_ : ∀ i, Nonempty (V i)), Nonempty (G ≃g (completeMultipartiteGraph V)) := by
+  constructor <;> intro h
+  · exact ⟨_, _, fun _ ↦ ⟨_, h.setoid.refl _⟩, ⟨h.iso⟩⟩
+  · obtain ⟨_, _, _, ⟨e⟩⟩ := h
+    intro _ _ _ h1 h2
+    rw [← e.map_rel_iff] at *
+    exact completeMultipartiteGraph.isCompleteMultipartite _ h1 h2
+
+lemma IsCompleteMultipartite.colorable_of_cliqueFree {n : ℕ} (h : G.IsCompleteMultipartite)
+    (hc : G.CliqueFree n) : G.Colorable (n - 1) :=
+    (completeMultipartiteGraph.colorable_of_cliqueFree _ (fun _ ↦ ⟨_, h.setoid.refl _⟩)
+          <| hc.comap h.iso.symm.toEmbedding).of_embedding h.iso.toEmbedding
+
+variable (G) in
+/--
+The vertices `v, w₁, w₂` form an `IsPathGraph3Compl` in `G` iff `w₁w₂` is the only edge present
+between these three vertices. It is a witness to the non-complete-multipartite-ness of `G`
```

Comment:
```suggestion
The vertices `v, w₁, w₂` form an `IsPathGraph3Compl` in `G` iff `w₁w₂` is the only edge present
between these three vertices. It is a witness to the non-complete-multipartite-ness of `G`.
This structure is an explicit way of describing that induced graph on `{v, w₁, w₂}` is the complement of `P3`.
```

Reply:
(or similar, to explain the name)

Reply:
Also let's maybe explicitly reference `not_isCompleteMultipartite_iff_exists_isPathGraph3Compl` here when you talk about the witness bit